### ISSUE TITLE
[feature] Support rules that don't specify `kind` or `pattern`

### ISSUE THREAD

**Issue by StevenLove** [ast-grep/ast-grep#1226]
# Problem
Sometimes we are trying to match a node or group of nodes primarily due to their relation to other nodes. For instance, we could want to match every node in a statement block that follows an important comment like `// DELETE BELOW`.
However, if we make a rule like
```yml
inside: {kind: statement_block}
follows: {kind: line_comment, regex: DELETE BELOW}
```
that does not have a `kind` or `pattern` then we typically get a `⚠ RuntimeError: unreachable`. 

# ⭐ Suggestion
Allow a rule to match any kind of node if `kind` is unspecified

# 💻 Use Cases

- [Matching all of the nodes in a statement block that occur after a specific comment](https://discord.com/channels/1107749847722889217/1191520520726913155/1252194080729858110)
- [Matching top-level nodes in a file](https://discord.com/channels/1107749847722889217/1245356688366764152/1245633411838906368)

# Workaround
One workaround is to use `any:[kind: kind1, kind: kind2 ...]` but it can be hard to enumerate every possible kind. If we had a way to say `kind: any` then this would be easy.

**Comment by HerringtonDarkholme** [comment]
Thanks for the suggestion. However, the requirement of `kind` is by design. ast-grep rules are grouped together by their kinds in order to skip matching rules against every node.

In your example, 
 
```yaml
inside: {kind: statement_block}
follows: {kind: line_comment, regex: DELETE BELOW}
```

Inside and follows have O(n2) time complexity and they can be very slow if matched against lots of nodes.

I would recommend using API is better here.

**Comment by StevenLove** [comment]
Good point. I agree that this feature would make it easy to write a very slow rule. I still believe that there are good use cases for this feature that make it outweigh the negative. I look forward to `fix` being implemented in the API (#1172) so that we can have a complete workaround. Thank you so much for your hard work!

**Comment by memel3** [comment]
inside: {kind: statement_block}
follows: {kind: line_comment, regex: DELETE BELOW}
